lib/linked_list.rb

```diff
@@ -18,95 +18,186 @@ def initialize
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+      new_node = Node.new(value, @head)
+      @head = new_node
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n) - Must traverse list
+    # Space Complexity: O(1)
     def search(value)
-      raise NotImplementedError
+      current = @head
+      while current
+        return true if value == current.data
+        current = current.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n) - Must traverse list
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+      current = @head
+      max = current&.data

```

Comment:
Why the `&`??

Reply:
This is to avoid an error stating that there is no such method .data for nil